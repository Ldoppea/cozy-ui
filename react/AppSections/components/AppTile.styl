.AppTile-icon-wrapper
  margin-top .25rem
  margin-bottom .25rem
  width 3rem
  height 3rem


.AppTile-icon-wrapper .AppTile-icon
  object-fit contain
  max-width 100%
  width auto


.AppTile-icon--default
  height 100%
  padding 0 .5em


.AppTile-desc
  display flex
  flex-direction column
  justify-content flex-start
  align-items center
  margin-top .5rem
  text-align center


.AppTile-title, .AppTile-developer, .AppTile-status
  display block
  width var(--item-size)
  margin 0
  text-overflow ellipsis
  white-space nowrap
  overflow hidden


.AppTile-title
  height 1.25rem


.AppTile-developer
  height 1rem
  font-size 12px
  color var(--slateGrey)


.AppTile-status
  margin-top .5rem
  height 1rem
  font-size 12px
  color var(--coolGrey)


@media (max-width 48rem)
  .AppTile-icon-wrapper
    margin-bottom 0

  .AppTile-icon
    height 2.5rem
    max-width 2.5rem
    max-height 2.5rem


  .AppTile-desc
    margin-top .125rem
    margin-left .5rem
    justify-content center
    text-align left
    /* 4rem for icon and margins */
    width calc(100% - 4rem)


  .AppTile-title
    height 1rem
    font-weight normal
    font-size 12px


  .AppTile-title, .AppTile-developer, .AppTile-status
    width 100%


  .AppTile-developer
    display none


  .AppTile-status
    margin-top .125rem
