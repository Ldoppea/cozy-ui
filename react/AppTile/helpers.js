export const getCurrentStatusLabel = app => {
  if (hasPendingUpdate(app)) return 'update'
  if (isUnderMaintenance(app)) return 'maintenance'
  if (isInstalledAndNothingToReport(app)) return 'installed'
  return null
}

export const hasPendingUpdate = app => {
  return !!app.availableVersion
}

export const isUnderMaintenance = app => {
  return !hasPendingUpdate(app) && app.maintenance
}

/* installed here means no actions needed */
export const isInstalledAndNothingToReport = app => {
  return !hasPendingUpdate(app) && !isUnderMaintenance(app) && app.installed
}
