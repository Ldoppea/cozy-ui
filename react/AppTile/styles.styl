@require 'settings/breakpoints.styl'

.AppTile-icon-wrapper
  margin-top .25rem
  margin-bottom .25rem
  width 3rem
  height 3rem

.AppTile-icon-wrapper .AppTile-icon
  object-fit contain
  max-width 100%
  width auto

.AppTile-icon--default
  height 100%
  padding 0 .5em

.AppTile-desc
  display flex
  flex-direction column
  justify-content flex-start
  align-items center
  margin-top .5rem
  text-align center
  max-width 100%

.AppTile-title, .AppTile-developer, .AppTile-status
  display block
  margin 0
  text-overflow ellipsis
  white-space nowrap
  overflow hidden

.AppTile-title
  height 1.25rem
  color var(--black)

.AppTile-developer
  height 1rem

.AppTile-status
  margin-top .5rem
  height 1rem
  color var(--coolGrey)


+small-screen()
  .AppTile-icon-wrapper
    margin-bottom 0

  .AppTile-icon
    height 2.5rem
    max-width 2.5rem
    max-height 2.5rem

  .AppTile-desc
    margin-top .125rem
    margin-left .5rem
    justify-content center
    text-align left
    /* 4rem for icon and margins */
    width calc(100% - 4rem)

  .AppTile-title
    height 1rem
    font-weight normal
    font-size 12px

  .AppTile-title, .AppTile-developer, .AppTile-status
    width 100%

  .AppTile-developer
    display none

  .AppTile-status
    margin-top .125rem
