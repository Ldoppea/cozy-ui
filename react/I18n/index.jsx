/**
 * Provides an I18n helper using a Higher Order Component.
 */

'use strict'

import React, { Component, useContext } from 'react'
import PropTypes from 'prop-types'

import { initTranslation } from './translation'
import { initFormat } from './format'

export const DEFAULT_LANG = 'en'

export const I18nContext = React.createContext()

// Provider root component
export class I18n extends Component {
  constructor(props) {
    super(props)
    this.init()
    this.state = {
      contextValue: this.getContextValue()
    }
  }

  init() {
    const { polyglot, lang, dictRequire, context, defaultLang } = this.props
    this.translator =
      polyglot || initTranslation(lang, dictRequire, context, defaultLang)
    this.format = initFormat(lang, defaultLang)
    this.t = this.translator.t.bind(this.translator)
  }

  getContextValue() {
    return {
      t: this.t,
      f: this.format,
      lang: this.props.lang
    }
  }

  getChildContext() {
    return this.state.contextValue
  }

  componentDidUpdate(prevProps) {
    if (prevProps.lang !== this.props.lang) {
      this.init()
      this.setState({ contextValue: this.getContextValue() })
    }
  }

  render() {
    return (
      <I18nContext.Provider value={this.state.contextValue}>
        {this.props.children}
      </I18nContext.Provider>
    )
  }
}

I18n.propTypes = {
  lang: PropTypes.string.isRequired, // current language.
  polyglot: PropTypes.object, // A polyglot instance.
  dictRequire: PropTypes.func, // A callback to load locales.
  context: PropTypes.string, // current context.
  defaultLang: PropTypes.string // default language. By default is 'en'
}

I18n.defaultProps = {
  defaultLang: DEFAULT_LANG
}

I18n.childContextTypes = {
  t: PropTypes.func,
  f: PropTypes.func,
  lang: PropTypes.string
}

// higher order decorator for components that need `t` and/or `f`
export const translate = () => WrappedComponent => {
  const Wrapper = props => {
    const i18nContext = useContext(I18nContext)
    return (
      <WrappedComponent
        {...props}
        t={i18nContext && i18nContext.t}
        f={i18nContext && i18nContext.f}
        lang={i18nContext && i18nContext.lang}
      />
    )
  }
  Wrapper.displayName = `withI18n(${WrappedComponent.displayName ||
    WrappedComponent.name})`
  return Wrapper
}

export const useI18n = () => useContext(I18nContext)

export { initTranslation, extend } from './translation'

export default I18n
