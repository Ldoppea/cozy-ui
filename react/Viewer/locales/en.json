{
  "Viewer": {
    "download": "Download",
    "openWith": "Open with...",
    "openInOnlyOffice": "Open with Only Office",
    "retry": "Retry",
    "error": {
      "network": "This file could not be loaded. Do you have a working internet connection right now?",
      "noapp": "No application on your device can open this file with mime type: %{fileMime} ",
      "missing": "This file is missing",
      "offline": "You should be connected to open this file"
    },
    "close": "close",
    "goto": "Go to %{url}",
    "next": "Next",
    "previous": "Previous",
    "scaledown": "Zoom out",
    "scaleup": "Zoom in"
  }
}
