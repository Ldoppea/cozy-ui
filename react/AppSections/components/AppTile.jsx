import React from 'react'
import PropTypes from 'prop-types'
import cx from 'classnames'

import AppIcon from '../../AppIcon'
import Tile from '../../Tile'
import { useI18n } from '../../I18n'
import { Caption } from '../../Text'

import { getCurrentStatusLabel } from '../status'
import { AppDoctype } from '../../proptypes'
import styles from './AppTile.styl'
let dataset
const getDataset = () => {
  if (dataset) return dataset
  const root = document.querySelector('[role=application]')
  dataset = root && root.dataset
  return dataset
}

const getAppIconProps = () => ({
  domain: getDataset() && getDataset().cozyDomain,
  secure: window.location.protocol === 'https:'
})

export const AppTitle = ({ children }) => (
  <h4 className={styles['AppTile-title']}>{children}</h4>
)

export const AppDeveloper = ({ children }) => (
  <Caption className={cx(styles['AppTile-developer'], 'u-slateGrey')}>
    {children}
  </Caption>
)

export const AppStatus = ({ children }) => (
  <Caption className={styles['AppTile-status']}>{children}</Caption>
)

export const AppTile = ({
  app,
  name,
  namePrefix,
  onClick,
  showDeveloper,
  IconComponent
}) => {
  const { t } = useI18n()
  const { developer = {} } = app
  const statusToDisplay = getCurrentStatusLabel(app)
  IconComponent = IconComponent || AppIcon
  return (
    <Tile tag="button" type="button" onClick={onClick}>
      <div className={styles['AppTile-icon-wrapper']}>
        <IconComponent
          app={app}
          className={styles['AppTile-icon']}
          {...getAppIconProps()}
        />
      </div>
      <div className={styles['AppTile-desc']}>
        <AppTitle>{namePrefix ? `${namePrefix} ${name}` : name}</AppTitle>
        {developer.name && showDeveloper && (
          <AppDeveloper>{`${t('app_item.by')} ${developer.name}`}</AppDeveloper>
        )}
        {statusToDisplay && (
          <AppStatus>{t(`app_item.${statusToDisplay}`)}</AppStatus>
        )}
      </div>
    </Tile>
  )
}

AppTile.propTypes = {
  app: AppDoctype,
  name: PropTypes.string.isRequired,
  namePrefix: PropTypes.string,
  onClick: PropTypes.func,
  showDeveloper: PropTypes.bool
}

AppTile.defaultProps = {
  showDeveloper: true
}

export default AppTile
