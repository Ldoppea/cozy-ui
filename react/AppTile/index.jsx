import React from 'react'

import AppIcon from '../AppIcon'
import PropTypes from 'prop-types'
import Tile, {
  TileTitle,
  TileSubtitle,
  TileFooter,
  TileIcon,
  TileDescription
} from '../Tile'

import { useI18n } from '../I18n'
import withLocales from '../I18n/withLocales'
import { AppDoctype } from '../proptypes'

import { getCurrentStatusLabel } from './helpers'
import styles from './styles.styl'
import en from './locales/en.json'
import fr from './locales/fr.json'

const locales = { en, fr }

let dataset
const getDataset = () => {
  if (dataset) return dataset
  const root = document.querySelector('[role=application]')
  dataset = root && root.dataset
  return dataset
}

const getAppIconProps = () => ({
  domain: getDataset() && getDataset().cozyDomain,
  secure: window.location.protocol === 'https:'
})

export const AppTile = ({
  app,
  name,
  namePrefix,
  onClick,
  showDeveloper,
  IconComponent
}) => {
  const { t } = useI18n()
  const { developer = {} } = app
  const statusToDisplay = getCurrentStatusLabel(app)
  IconComponent = IconComponent || AppIcon
  return (
    <Tile tag="button" type="button" onClick={onClick}>
      <TileIcon>
        <IconComponent
          app={app}
          className={styles['AppTile-icon']}
          {...getAppIconProps()}
        />
      </TileIcon>
      <TileDescription>
        <TileTitle>{namePrefix ? `${namePrefix} ${name}` : name}</TileTitle>
        {developer.name && showDeveloper && (
          <TileSubtitle>{`${t('app_item.by')} ${developer.name}`}</TileSubtitle>
        )}
        {statusToDisplay && (
          <TileFooter>{t(`app_item.${statusToDisplay}`)}</TileFooter>
        )}
      </TileDescription>
    </Tile>
  )
}

AppTile.propTypes = {
  app: AppDoctype,
  name: PropTypes.string.isRequired,
  namePrefix: PropTypes.string,
  onClick: PropTypes.func,
  showDeveloper: PropTypes.bool
}

AppTile.defaultProps = {
  showDeveloper: true
}

export default withLocales(locales)(AppTile)
