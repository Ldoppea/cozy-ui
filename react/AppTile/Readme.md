Component used to show an app status, can be used in
a list of apps for example in the Store.

```
import AppTile from '.'
import mockApps from '../AppSections/_mockApps'
import { I18n } from '../I18n'

const locale = {}

const app = mockApps[0];

<I18n dictRequire={lang => locale} lang="en">
  <div className='u-flex'>
    <AppTile app={app} />
    <AppTile app={{...app, maintenance: true}} />
    <AppTile app={{...app, availableVersion: true}} />
    <AppTile app={{...app, availableVersion: true}} showStatus={['maintenance']} />
  </div>
</I18n>
```
