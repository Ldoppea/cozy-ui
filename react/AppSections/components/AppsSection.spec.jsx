'use strict'

/* eslint-env jest */

import React from 'react'
import { mount } from 'enzyme'

import { CozyProvider } from 'cozy-client'
import { I18nContext } from '../../jestLib/I18n'
import { AppsSection } from './AppsSection'
import AppTile from './AppTile'

import mockApps from '../../mocks/apps'
import en from '../locales/en'
import { I18n } from '../../I18n'

const i18nContext = I18nContext({ locale: en })
const tMock = i18nContext.t

describe('AppsSection component', () => {
  let component
  const client = {}
  const setup = ({ onAppClick }) => {
    component = mount(
      <CozyProvider client={client}>
        <I18n lang="en" dictRequire={() => en}>
          <AppsSection
            t={tMock}
            lang="en"
            subtitle={<h3>Test Apps</h3>}
            appsList={mockApps}
            onAppClick={onAppClick}
          />
        </I18n>
      </CozyProvider>
    ).find(AppsSection)
  }

  it('should be rendered correctly with apps list, subtitle', () => {
    const mockOnAppClick = jest.fn()
    setup({ onAppClick: mockOnAppClick })
    expect(component.find('h3').map(x => x.text())).toMatchSnapshot()
    expect(component.find('h4').map(x => x.text())).toMatchSnapshot()
    expect(component.find('p').map(x => x.text())).toMatchSnapshot()
  })

  it('should run provided onAppClick on AppTile click event', () => {
    const mockOnAppClick = jest.fn()
    setup({ onAppClick: mockOnAppClick })
    expect(component.find(AppTile).length).toBe(mockApps.length)
    const appItem = component.find(AppTile).at(0)
    appItem.simulate('click')
    expect(mockOnAppClick.mock.calls.length).toBe(1)
    expect(mockOnAppClick.mock.calls[0][0]).toBe('konnector-bouilligue')
  })
})
