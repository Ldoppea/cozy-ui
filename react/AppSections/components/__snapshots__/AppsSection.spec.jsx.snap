// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppsSection component should be rendered correctly with apps list, subtitle 1`] = `
Array [
  "Test Apps",
]
`;

exports[`AppsSection component should be rendered correctly with apps list, subtitle 2`] = `
Array [
  "Bouilligue",
  "Cozy Collect",
  "DevOnly",
  "Cozy Drive",
  "Cozy Photos",
  "Tasky",
  "Trinlane",
]
`;

exports[`AppsSection component should be rendered correctly with apps list, subtitle 3`] = `
Array [
  "By Cozy",
  "By Cozy",
  "Installed",
  "By Cozy",
  "By Cozy",
  "Installed",
  "By Cozy",
  "Update available",
  "By Cozy",
  "Installed",
]
`;
